### PR TITLE
Added security context for system collections

### LINKED ISSUE #2038: system collection should have their own security context
System collections should have their own security context, because they are not allowed to be deleted at all (We can make use of the new feature introduced in https://github.com/sulu-io/sulu/pull/2036).

Required because the admin wants to be able to hide the system collections from specific roles/users.

### PR DESCRIPTION
**Pull request by danrot**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes #2038 
| Related issues/PRs | https://github.com/massiveart/husky/pull/623
| License | MIT
| Documentation PR | none

#### What's in this PR?

This PR adds a new security context called `sulu.media.system_collections`, which only allows to set the `view` permission. Based on this permission system collections and its containing assets are only displayed if the system grants this permission to the logged in user (in both, the collections administration and the media selection content type).

### CODE REVIEW

**Comment by danrot** [comment]
IMO it's not optimal that all the repositories and managers in the media bundle differ quite a lot regarding the API. Some of them have a `filter` array, some have separate parameters instead. Sometimes the limit comes first and sometimes the offset. However, I left all these things, since it would be out of scope of this PR and a huge BC break to change that.

File: src/Sulu/Bundle/MediaBundle/Controller/MediaController.php [hunk 1]
```diff
@@ -74,7 +76,13 @@ public function getAction($id, Request $request)
             $view = $this->responseGetById(
                 $id,
                 function ($id) use ($locale, $mediaManager) {
-                    return $mediaManager->getById($id, $locale);
+                    $media = $mediaManager->getById($id, $locale);
+
+                    if ($media->getEntity()->getCollection()->getType()->getKey() === 'collection.system') {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaController.php]: Isn't the `collection.system` string already stored somewhere? I am already duplicating it...

Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaController.php]: @wachterjohannes Isn't the `collection.system` string already stored somewhere? I am already duplicating this literal...

File: src/Sulu/Bundle/MediaBundle/Controller/MediaController.php [hunk 2]
```diff
@@ -96,10 +104,18 @@ function ($id) use ($locale, $mediaManager) {
     public function cgetAction(Request $request)
     {
         try {
+            $collectionId = $request->get('collection');
+
+            if ($collectionId) {
+                $collection = $this->getCollectionManager()->getById($collectionId, $this->getLocale($request));
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaController.php]: I don't know a different way than loading the collection here, in order to be able to check if it is a system collection, and check if the user is allowed to see the content... /cc @chirimoya

File: src/Sulu/Bundle/MediaBundle/Entity/MediaRepository.php [hunk 3]
```diff
@@ -274,12 +304,18 @@ private function getIdsQuery(
...
+
         if (!empty($collection)) {
-            $queryBuilder->innerJoin('media.collection', 'collection');
             $queryBuilder->andWhere('collection.id = :collection');
             $queryBuilder->setParameter('collection', $collection);
         }
 
+        if (!$systemCollections) {
+            $queryBuilder->leftJoin('collection.type', 'type');
+            $queryBuilder->andWhere('type.key != \'collection.system\'');
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Entity/MediaRepository.php]: That's where this literal is duplicated...

**Comment by danrot** [comment]
@chirimoya @wachterjohannes Would be awesome if you could have a quick look if that is ok for you. Turned out it is not too easy to add the tests here, so I would like to know if what I am doing is ok beforehand.

File: src/Sulu/Bundle/MediaBundle/Controller/MediaController.php (hunk 1 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaController.php]: i think there is a db entry but no const if you thought about that!

File: src/Sulu/Bundle/MediaBundle/Controller/MediaController.php [hunk 4]
```diff
@@ -74,7 +76,13 @@ public function getAction($id, Request $request)
             $view = $this->responseGetById(
                 $id,
                 function ($id) use ($locale, $mediaManager) {
-                    return $mediaManager->getById($id, $locale);
+                    $media = $mediaManager->getById($id, $locale);
+
+                    if ($media->getEntity()->getCollection()->getType()->getKey() === 'collection.system') {
+                        $this->getSecurityChecker()->checkPermission('sulu.media.system_collections', 'view');
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaController.php]: use permissiontype::view

File: src/Sulu/Bundle/MediaBundle/Controller/MediaController.php (hunk 1 above)
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaController.php]: Then I should probably introduce something like that... The question is where :confused:

Comment (wachterjohannes) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaController.php]: i would place it into CollectionManager or SystemCollectionManager

File: src/Sulu/Bundle/MediaBundle/Controller/MediaController.php (hunk 2 above)
Comment (chirimoya) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaController.php]: ACK

**Pull request merged**